File: Leetcode/DP/Out_of_Boundary_Paths.py
class Solution:
    
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        dr = [1,0,-1,0]
        dc = [0,1,0,-1]
        mod = 10**9 +7

        @lru_cache(None)
        def dfs(M, N, r, c, movesLeft):
            if r<0 or r>=M or c<0 or c>=N:
                return 1
            if movesLeft==0:
                return 0

            res = 0
            for i in range(4):
                res+=dfs(M,N,r+dr[i], c+dc[i], movesLeft-1)%mod
            return res%mod
        
        return dfs(m,n,startRow, startColumn,maxMove)
        

File: Leetcode/DP/Out_of_Boundary_Paths.java
class Solution
{
    Long dp[][][]; // dp[x][y][maxMoves];
    int mod = (int)1e9+7;
    int dr[] = {1,0,-1,0,1};
    public long count(int x, int y, int maxMove, int m, int n)
    {
        if(Math.min(x,y)<0 || x>=m || y>=n) return 1;
        if(dp[x][y][maxMove] != null) return dp[x][y][maxMove];
        long cnt = 0;
        for(int i = 0; i<4; i++)
        {
            int nx = x+dr[i], ny = y+dr[i+1];
            if(maxMove>0) 
            {
                cnt+=count(nx,ny, maxMove-1,m,n);
                cnt%=mod;
            }
        }
        return dp[x][y][maxMove] = cnt;
        
    }
    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        dp = new Long[m][n][maxMove+1];
        return (int) count(startRow, startColumn, maxMove, m,n);
    }
}